mackerel-plugin-openldap/lib/openldap.go

```diff
@@ -0,0 +1,355 @@
+	}
+	return res, errors.New("not found CSN")
+}
+
+func getLatestCSN(host, base, bind, passwd string, useTLS, insecureSkipVerify bool) (time.Time, error) {
+	var l *ldap.Conn
+	var err error
+	var res time.Time
+	if useTLS {
+		l, err = ldap.DialTLS("tcp", host, &tls.Config{InsecureSkipVerify: insecureSkipVerify})
+	} else {
+		l, err = ldap.Dial("tcp", host)
+	}
+	err = l.Bind(bind, passwd)
+	if err != nil {
+		logger.Errorf("Failed to Bind %s, err: %s", bind, err)
+		return res, err
+	}
+	searchRequest := ldap.NewSearchRequest(base, ldap.ScopeBaseObject, ldap.NeverDerefAliases, 0, 0, false, "(&(objectClass=*))", []string{"ContextCSN"}, nil)
+	sr, err := l.Search(searchRequest)
+	l.Close()
+	if err != nil {
+		logger.Errorf("Failed to ldap search. %s.", err)
+		return res, err
+	}
+	return latestCSN(sr)
+
+}
+
+// FetchMetrics interface for mackerelplugin
+func (m OpenLDAPPlugin) FetchMetrics() (map[string]interface{}, error) {
+	stat := make(map[string]float64)
+	if m.ReplBase != "" {
+		masterTime, err := getLatestCSN(m.ReplMasterHost, m.ReplBase, m.ReplMasterBind, m.ReplMasterPass, m.ReplMasterUseTLS, m.InsecureSkipVerify)
+		if err != nil {
+			return nil, err
+		}
+		localTime, err := getLatestCSN(m.TargetHost, m.ReplBase, m.ReplLocalBind, m.ReplLocalPass, m.UseTLS, m.InsecureSkipVerify)
+		if err != nil {
+			return nil, err
+		}
+		stat["replication_delay"] = float64(masterTime.UnixNano() - localTime.UnixNano())
+	}
+
+	ldapOpes, err := fetchOpenldapMetrics(m.l, "cn=Operations,cn=Monitor", "", []string{"monitorOpInitiated", "monitorOpCompleted"})
+	if err != nil {
+		return nil, err
+	}
+	mergeStat(stat, ldapOpes)
+	ldapWaiters, err := fetchOpenldapMetrics(m.l, "cn=Waiters,cn=Monitor", "waiters_", []string{"monitorCounter"})
+	if err != nil {
+		return nil, err
+	}
+	mergeStat(stat, ldapWaiters)
+
+	ldapMaxThreads, err := fetchOpenldapMetrics(m.l, "cn=Max,cn=Threads,cn=Monitor", "threads_", []string{"monitoredInfo"})
+	if err != nil {
+		return nil, err
+	}
+	mergeStat(stat, ldapMaxThreads)
+	ldapOpenThreads, err := fetchOpenldapMetrics(m.l, "cn=Open,cn=Threads,cn=Monitor", "threads_", []string{"monitoredInfo"})
+	if err != nil {
+		return nil, err
+	}
+	mergeStat(stat, ldapOpenThreads)
+	ldapActiveThreads, err := fetchOpenldapMetrics(m.l, "cn=Active,cn=Threads,cn=Monitor", "threads_", []string{"monitoredInfo"})
+	if err != nil {
+		return nil, err
+	}
+	mergeStat(stat, ldapActiveThreads)
+
+	ldapStatistics, err := fetchOpenldapMetrics(m.l, "cn=Statistics,cn=Monitor", "statistics_", []string{"monitorCounter"})
+	if err != nil {
+		return nil, err
+	}
+	mergeStat(stat, ldapStatistics)
+	ldapTotalConns, err := fetchOpenldapMetrics(m.l, "cn=Total,cn=Connections,cn=Monitor", "connections_", []string{"monitorCounter"})
+	if err != nil {
+		return nil, err
+	}
+	mergeStat(stat, ldapTotalConns)
+	ldapCurrentConns, err := fetchOpenldapMetrics(m.l, "cn=Current,cn=Connections,cn=Monitor", "connections_", []string{"monitorCounter"})
```

Comment:
[memo]
Those connection metrics seems to be gathered by 1 search query `cn=Connections,cn=Monitor`, but this query also corrects stats about all current connections, so this 2-query-style is better. 🙆‍♂️